include/cheri/mode/cheri.h

```diff
@@ -0,0 +1,60 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2019 Hesham Almatary
+ *
+ * This software was developed by SRI International and the University of
+ * Cambridge Computer Laboratory (Department of Computer Science and
+ * Technology) under DARPA contract HR0011-18-C-0016 ("ECATS"), as part of the
+ * DARPA SSITH research programme.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *
+ * @brief CHERI utility functions
+ */
+
+#pragma once
+
+#include <types.h>
+#include <linker.h>
+
+typedef __SIZE_TYPE__ size_t;
+typedef __SIZE_TYPE__ ptraddr_t;
+
+void *__capability cheri_seal_cap(void *__capability unsealed_cap, size_t otype);
+void *__capability cheri_unseal_cap(void *__capability unsealed_cap);
```

Comment:
How is authority for capability sealing and unsealing expected to be managed in seL4? What authority is needed to be able to seal a capability?

> Capability sealing The CSeal and CUnseal instructions seal or unseal capabilities given a
suitable authorizing capability. Sealed capabilities allow software to implement encap-
sulation, such as is required for software compartmentalization.

Reply:
> How is authority for capability sealing and unsealing expected to be managed in seL4? What authority is needed to be able to seal a capability?
> 
It currently isn't expected to be used in seL4 kernel itself (similar to CheriBSD). But it's just provided as CHERI helpers here in case we do in the future. The user, however, can use sealing as they wish (and there are many user-level compartmentalisation use-cases for it). 

To answer your question, in order to seal/unseal a capability, you'd have to have a "sealer/unsealer" capability that:
1. Has sealing/unsealing permissions.
2. Has the same `otype` as the capability you're sealing/unsealing. 

> > Capability sealing The CSeal and CUnseal instructions seal or unseal capabilities given a
> > suitable authorizing capability. Sealed capabilities allow software to implement encap-
> > sulation, such as is required for software compartmentalization.

Reply:
I probably need to look more at some examples.  Does a CHERI cap with sealing/unsealing permissions and no `otype` set represent authority to craft sealing/unsealing caps for any `otype`? Is that something that is controlled by the OS in cheriBSD or would it be given to the program's runtime? Thinking about how this could work on seL4. I think it's only relevant at this stage if there's suggestions of using sealing/unsealing to transfer CHERI pointers via IPC buffers.

Reply:
> I probably need to look more at some examples. Does a CHERI cap with sealing/unsealing permissions and no otype set represent authority to craft sealing/unsealing caps for any otype?
Yes

> Is that something that is controlled by the OS in cheriBSD or would it be given to the program's runtime? 
It's given to the program. CheriBSD doesn't manage or use that. 

> Thinking about how this could work on seL4. I think it's only relevant at this stage if there's suggestions of using sealing/unsealing to transfer CHERI pointers via IPC buffers.

Sealing and their invocations is still pretty much researchy. CheriBSD just lets the user do what they want with it, but we don't have production-ready uses for it AFAIK, and it's not used by default apart from sentries (just a special case for CFI). The sealing mechanism itself (e.g., otypes, invocations, unsealing, etc) also differs between Morello, our CHERI-RISC-V, CHERIoT, and the to-be-standardised [CHERI-RISC-V draft](http://github.com/riscv/riscv-cheri). I wouldn't strongly bound this PR to it, but we could keep thinking about solutions using it for future use, and probably make suggestions to something like CHERI-RISC-V ISA draft to better accommodate CHERI-seL4 needs if we think of any.